auacm/angular/src/app/view-scoreboard/view-scoreboard.component.ts

```diff
@@ -0,0 +1,180 @@
+        this._competitionService.fetchCompetition(this.competition.cid);
+      }
+    })
+  }
+
+  ngOnInit() {
+    this.userData = this._userService.getUserData();
+    this.clientTimeOffset = this._competitionService.getClientTimeOffset();
+    this._activeRoute.params.subscribe(params => {
+      if (params['cid']) {
+        this._competitionService.fetchCompetition(+params['cid']);
+      } else {
+        this._router.navigate(['404']);
+      }
+    });
+  }
+
+  ngOnDestroy() {
+    // You need to clear the timer when the component is destroyed 
+    // otherwise the client will build up 1 billion timers. xD
+    if (this.scoreboardTimer !== undefined && this.scoreboardTimer !== null) {
+      for (var timer of this.scoreboardTimer) {
+        clearInterval(timer);
+      }
+    }
+    // If the user is on this page, clear the scoreboard data
+    this._competitionService.resetScoreboard();
+  }
+
+  refreshScoreboard() {
+    for (var team of this.competition.teams) {
+      var solved = 0;
+      var time = 0;
+      for (var key in team.problemData) {
+        if (team.problemData[key].status === 'correct') {
+          solved++;
+          team.problemData[key].penaltyTime = (team.problemData[key].submitCount - 1) * 20;
+          time += team.problemData[key].submitTime + team.problemData[key].penaltyTime;
+        } else {
+          team.problemData[key].penaltyTime = team.problemData[key].submitCount * 20
+        }
+      }
+      team.solved = solved;
+      team.time = time;
+    }
+
+    this.competition.teams.sort(function(a, b) {
+      if (a.solved != b.solved) {
+        return b.solved - a.solved;
+      } else {
+        return a.time - b.time;
+      }
+    });
+
+    if (this.competition.teams.length > 0) {
+      var rank = 1;
+      var prevSolved = this.competition.teams[0].solved;
+      var prevTime = this.competition.teams[0].time;
+      this.competition.teams[0].rank = rank;
+      for (var i = 1; i < this.competition.teams.length; i++) {
+        var team = this.competition.teams[i];
+        if (team.solved < prevSolved) {
+          rank++;
+          team.rank = rank;
+        } else if (team.solved == prevSolved && team.time > prevTime) {
+          rank++;
+          team.rank = rank;
+        } else {
+          team.rank = rank;
+        }
+        prevSolved = team.solved;
+        prevTime = team.time;
+      }
+    }
+
+    // Can break if an $apply is already in progress, need to check first (???)
+    // if (!$scope.$$phase) {
+    //     $scope.$digest();
+    // }
+  };
+
+  problemIsInComp(problemId: number): boolean {
+    for (var problem in this.competition.compProblems) {
+      if (problemId === this.competition.compProblems[problem].pid) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  calculateCompetitionProgress() {
+    if (this.scoreboardTimer.length < 1) {
+      var clientTime = Math.floor((Date.now() + this.clientTimeOffset) / 1000);
```

Comment:
It seems as though clientTime is actually referring to time according to the server. If this is the case, this variable name is misleading, IMO.